Guard

chiragsibm opened the pull request.

Commit:
libguard: Adding GardFlags and GardFlagPnorWriteInProgress enums

Change:
Adding GardFlags and GardFlagPnorWriteInProgress enums to handle
GUARD partition write.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

Commit:
libguard: Adding function to set/reset the write flag

Change:
Adding new function to set/reset the write flag.
While creating guard record or re-writing a guard record
write flag will be set to 0x00 and once the write completes,
set the flag to 0x01.

Tested:
While writing the guard data the flag is set as 0x00
and once the write is finished its set back as 0x01.
Created guard records to check if data is coming as
expected or not.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
     return pos;
```
Commit message:

Please elaborate on the problem and add test case details.

we could split this commit as 2 commit
- New flag and API (it can talk about the actual issue and solution) 
- libguard using that API to support BMC flow too (The reason for updating the flag in the libguard too)

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+ * @brief Function to set/reset write flag in guard header
+ *
+ * @param[in] GuardFile Guard file path
+ * @param[in] set - true if writing the guard record
```
JS: I am not sure what format we are using in the API documentation in this repo. Please keep consistent (looks like you added `-` for one parameter but not another one)

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+ *                  false to reset the guard write flag
+ *
+ */
+void toggleWriteFlag(GuardFile& file, bool set = false)
```
JS: `const bool set = false`?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+void toggleWriteFlag(GuardFile& file, bool set = false)
+{
+    GuardRecord_t guardRecord;
+    size_t headerPos = 9;
```
headerFlagPos?

JS: You could use `constexpr` since it won't be changed.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+    else
+    {
+        // Resetting write bit in guard header
+        guardRecord.iv_flags = GARD_FLAGS_DEFAULT;
```
I think we need set it as `0`, right? why `GARD_FLAGS_DEFAULT`?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+        // Resetting write bit in guard header
+        guardRecord.iv_flags = GARD_FLAGS_DEFAULT;
+        file.write(headerPos, &guardRecord.iv_flags,
+                   sizeof(guardRecord.iv_flags));
```
JS: you could move this before if and else block and use it.

RameshIyyar commented on libguard/include/guard_record.hpp:
```diff
@@ -14,6 +14,12 @@ namespace guard
 const uint8_t CURRENT_GARD_VERSION_LAYOUT = 0x1;
 #define GUARD_RESOLVED 0xFFFFFFFF
 
+enum GardFlags : uint8_t
```
Please align with the HB record format.

they are using bit7 from the flag (1byte) to indicate whether the write is in progress or not so, we should take care of the endianness.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -118,6 +118,34 @@ int guardNext(GuardFile& file, int pos, GuardRecord& guard)
+    if (set)
+    {
+        // Setting write bit in guard header
+        guardRecord.iv_flags = GARD_FLAG_MASK_PNOR_WRITE_IN_PROGRESS;
```
I think we will hit the endianness issue here. Can please confirm with the HB team?

BMC is little-endian and HB is big-endian (pnor), HB uses the 7th bit from the flag but, BMC will set in the 0th bit in the flag.

The same comment is applicable to the `else` block.

chiragsibm replied:
This is unit8_t no endianness issue for 1 byte data.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -158,6 +186,8 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType,
     memset(&existGuard, 0xff, sizeOfGuard);
 
     GuardFile file(guardFilePath);
+    toggleWriteFlag(file, true);
```
Why cannot just use it inside the GuardFile::write() instead of calling many places? any specific reason?

chiragsibm replied:
To keep the guard_file.cpp only do basic read and write operation, any external operation should be done in interface.cpp i.e. writing to header etc.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -426,6 +426,20 @@ void invalidateAll()
+{
+    GuardFile file(guardFilePath);
+    GuardRecord_t guardRecord;
+    size_t headerPos = 9;
```
headerFlagPos?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -426,6 +426,20 @@ void invalidateAll()
+    size_t headerPos = 9;
+    file.read(headerPos, &guardRecord.iv_flags, sizeof(guardRecord.iv_flags));
+
+    if (guardRecord.iv_flags == 0x00)
```
I think we should check specific bit, please take care endianness.

Commit:
libguard: Adding function to check write flag

Change:
Adding function checkWriteflag, which will read the
write flag.

Tested:
In ipl code calling this function and its working
as expected i.e. if flag is set to 0x00 return true
else false.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

devenrao commented on libguard/guard_interface.cpp:
```diff
@@ -333,10 +372,12 @@ static void invalidateRecord(const guardRecordParam& value)
              (existGuard.targetId == entityPath)) &&
             (existGuard.recordId != GUARD_RESOLVED))
         {
+            toggleWriteFlag(file);
```
this is not correct, these chain operatoins are dangerous you look for CustoFD class you need to follow that, constructor of CustomFd opens the file and destructor closes the fd. 

{
CustomFD fd;
.
.
}
when it reaches here fd will be closed

ojayanth commented:
Please include summary in the PR why these chnages introduced.  PR name  "Huard" is not making any sense here.
